### PR TITLE
feat: add dense spherical harmonic engine

### PR DESCRIPTION
**Pull request by slosar**
## Summary

- add an optional `engine="dense"` transform path for repeated low-band-limit workloads
- generate and cache packed `m >= 0` analysis matrices, then apply them with native JAX matrix multiplication
- preserve JIT compilation, batching, reverse-mode differentiation, and finite `niter` refinement
- allow HEALPix `Beam` and `Sky` objects to select an explicit `lmax` below the default `2 * nside`
- expose matrix precomputation and cache-clearing helpers and document the new API

The existing `engine="s2fft"` behavior remains the default. The dense engine does not add an Almond, CuPy, callback, or JAXbind dependency.

For the motivating `nside=32, lmax=30` case, the packed complex128 matrix has shape `(496, 12288)` and occupies exactly 93 MiB. It built in about 1.44 seconds in a CPU-only local environment.

## Validation

- `uv run pytest -q` — 441 passed
- `uv run ruff check src tests`
- `uv run ruff format --check src tests`
- dense/s2fft equivalence through `niter=3` at approximately `1e-15` where both engines support the geometry
- gradient, enclosing-JIT, cache reuse, non-HEALPix compatibility, and low-`L`/high-`nside` tests included

GPU execution uses native JAX GEMM, but a CUDA-enabled JAX runtime was not available for local benchmarking.

### CODE REVIEW

File: src/croissant/sphere.py [hunk 1]
```diff
@@ -1,20 +1,339 @@
...
+    """Build a HEALPix matrix by evaluating spherical harmonics directly."""
+    from scipy import special
+
+    npix = int(np.prod(spatial_shape))
+    if spatial_shape != (npix,):
+        raise ValueError("HEALPix maps must have one spatial pixel axis")
+
+    ell, emm = _positive_lm_indices(lmax)
+    nalm = len(ell)
+    complex_dtype = jnp.complex128 if jax.config.x64_enabled else jnp.complex64
```
[the remark this answers is not in the document]
Comment (christianhbye) [reply src/croissant/sphere.py]: Verified empirically — the output dtype never diverges between engines. s2fft itself promotes float32 inputs to complex128 under an x64 runtime, so the flag-based choice here always lands on the same dtype s2fft produces:

| x64 flag | input maps | engine | dense matrix dtype | alm output dtype |
|---|---|---|---|---|
| enabled | float64 | s2fft | — | complex128 |
| enabled | float64 | dense | complex128 | complex128 |
| enabled | float32 | s2fft | — | complex128 |
| enabled | float32 | dense | complex128 | complex128 |
| disabled | any (coerced to float32) | s2fft | — | complex64 |
| disabled | float32 | dense | complex64 | complex64 |

The non-HEALPix builder path gives the same result (checked with `dh` sampling: float32 request still yields a complex128 matrix under x64), so both builders are consistent. Building at complex64 for float32 inputs would actually *create* the divergence this comment warns about, while saving memory only in a mode croissant already warns against at import.

The one real wart is that `dtype` participates in the cache key while not affecting the built matrix, so float32/float64 requests store duplicate identical matrices. We are considering simplifying the `dtype` argument out of the API to fix that.

[the remark this answers is not in the document]
Comment (christianhbye) [reply src/croissant/sphere.py]: @slosar It seems like the dtype argument does not do anything. Removing the argument and relying on x64 enabled/disabled is probably fine, or you can try to force the matrix to the requested dtype

[the remark this answers is not in the document]
Comment (slosar) [reply src/croissant/sphere.py]: OK, I think it should be fixed now.
